### PR TITLE
fix(tutor): tell user to add lines to vimtutor so the line checker doesn't break

### LINKED ISSUE #28164: tutor tests become unresponsive on line-number mismatch
### Problem
The line-gutter's test-decorators in `:Tutor` become unresponsive after their line numbers change i.e. due to removal/addition of lines before them.

while this is perfectly understandable: _of course, the test is line number specific"_
but the issue lies in the fact that the tutorial _demands_ lines to be added/deleted

i.e. there's a contradiction in how the tutorial interacts with users, vs how the tests assume it would stay.

### Minimal Steps to reproduce
<sub>I couldn't get the clipboard to work despite `scoop install win32yank` and restart terminal. So, some steps might feel convoluted.</sub>

**Normal conditions:**

0. `nvim -c ':Tutor'` or equivalently `nvim` > `:Tutor`
1. correct the L`333` 
(say by, `332gg` > paste the following correct line > `jdd`)
    - :-) see test pass responsively

**Trigger and buggy behaviour:**

2. `308gg` `dd` `332gg` or better yet, `kdd` (both are equivalent for our purpose)
3. `0i` `f` `<ESC>` (make any change to try to make test fail): 
    - :-( see that the `✓` stays there unresponsively

**Confirmation of the diagnosis/hypothesized cause:**

4. `ki` `<RET>` `<ESC>` to insert a blank line above (i.e. bring the test line number back to '333')
5. `j0a` `<SPC>` `<ESC>` (force refresh the test):
    - :-) see the test fail responsively
6. `0dw` (neutralise edits from step 5, 7):
    - :-) see the test pass responsively

```
Fix the errors on this line and replace them with undo.
```

### Expected behavior

the test indicator on line-gutter `✗✓` stays responsive during at least the normal flow of test...

[image](https://github.com/neovim/neovim/assets/19423063/ea2d586a-0a88-4185-ad19-88555bc48373)

### Solutions?

I am not sure what an ideal, MVC solution can be. Some thoughts come to mind:

**update:** it seems that the best and minimum viable solution can be this 2nd one:

2. provide a quick shortcut command which restarts the test afresh from the line the user is at at the time of invocation
    * would require: a reminder line in tutorial text (after a deleted line) reminding user to start the test afresh
    - no downsides as far as i can see
    - full flexibility
    - not much is required to change, all the functionality required to make it work already exists
    - i.e. true to `MVC`, simple, cost effective and elegant

[fold-marker]: https://neovim.io/doc/user/fold.html#fold-marker "fold-marker"

<details open><summary>alternate solutions considered:</summary>
<p>

1. _[hash-map + array][only-upfront-cost]_ define "named chunk ranges" somehow, & refer to line numbers relative to chunk's start in tests
   <sub>(dictionaries can solve everything eventually 🤣)</sub>
    + <sub>(inspired from `powershell` config)</sub>
    + neo/vim already has [marker-based folding range][fold-marker] which can fit this purpose very well
      nevertheless, it seems non-trivial to use that in this way
    + but once done; the rest will be easy ride
    + full flexibility of attempting tests in any order
	+ no constraints on what is done to lines outside of the test chunk: i.e. full freedom to experiment _and_ edit
        + no need to make sweeping changes to fix the line locations in tests on any line insertions in the tutor

^

0. _[brute-force][continous-cost]_ hard-coding the expected line numbers after the line deletions through test
    - _requires alll the tests_ and _nothing except the tests_ to be done _sequentially_
    - i.e. would also require to save the tutorial buffer, and resume from there the next time
    - overall, toooo much restrictive

</p>
</details> 

Sample from neo/vim's marker based folding range with `{{{<N> ... }}}<N>`: 

```markdown
<!-- Chapter 1. Navigation {{{3 -->


<!-- Chapter 2. Motions {{{3 -->
```

Sample from powershell's named chunks with `#region <name> ... #endregion` :

```powershell
#region conda initialize
# !! Contents within this block are managed by 'conda init' !!
If (Test-Path "...\mambaforge\Scripts\conda.exe") {
    (& "...\mambaforge\Scripts\conda.exe" "shell.powershell" "hook") | Out-String | ?{$_} | Invoke-Expression
}
#endregion
```

<details>
<summary><h2>Environment</h2></summary>
<p>

### Neovim version (nvim -v)

NVIM v0.9.5 Build type: Release LuaJIT 2.1.1710088188     system vimrc file: "$VIM\sysinit.vim"   fall-back for $VIM: "/mingw64/share/nvim"  Run :checkhealth for more info

### Vim (not Nvim) behaves the same?

not applicable

### Operating system/version

windows 10.0.19045 N/A Build 19045

### Terminal name/version

mintty 3.7.1 (x86_64-pc-msys)

### $TERM environment variable

xterm

### Installation

in msys2 `mingw-w64-x86_64-neovim` pkg via command: `pacboy -S neovim:x`

</p>
</details>

### PR DESCRIPTION
**Pull request by echometerain**
fixes #28164 

I am a novice vim user doing vimtutor on neovim. I have attempted it multiple times but I have always gotten stuck at the line `Fix the errors on this line and replace them with undo` because I was unaware that deleting lines, as per the instructions of `# Lesson 2.6: OPERATING ON LINES` which I just followed, destroys nvim's ability to check for the correctness of a line. It wasn't until I found #28164 and `runtime/tutor/en/vim-01-beginner.tutor.json` that I realized that `expect` operated on line numbers.

I think maybe we should tell novices what to do instead of letting them break it over and over again

### CODE REVIEW

**Comment by zeertzjq** [comment]
There is a problem: the user may not know how to do those from the information provided by the tutor so far.

File: runtime/tutor/en/vim-01-beginner.tutor [hunk 1]
```diff
@@ -302,7 +302,7 @@ it would be easier to simply type two d's to delete a line.
 
  3. Now move to the fourth line.
 
- 4. Type `2dd`{normal} to delete two lines.
+ 4. Type `2dd`{normal} to delete two lines, then add three lines back so the line numbers stay the same.
```
[the remark this answers is not in the document]
Comment (echometerain) [reply runtime/tutor/en/vim-01-beginner.tutor:305]: The user has already learned insert mode from earlier in the tutorial and `u` doesn't get taught until the next lesson. Maybe I'll modify it so the user knows to press enter 3 times in insert mode.

File: runtime/tutor/en/vim-01-beginner.tutor [hunk 2]
```diff
@@ -689,7 +689,7 @@ NOTE: Pressing [v](v) starts [Visual selection](visual-mode). You can move the c
  1. Place the cursor just above this line.
 
 NOTE: After executing Step 2 you will see text from Lesson 5.3. Then move
-      DOWN to see this lesson again.
+      DOWN to see this lesson again. Please undo after you are done.
```
Comment (justinmk) [inline runtime/tutor/en/vim-01-beginner.tutor:692]: do not use "please", it is noise.
```suggestion
      DOWN to see this lesson again. Then undo after you are done.
```

File: runtime/tutor/en/vim-01-beginner.tutor (hunk 1 above)
[the remark this answers is not in the document]
Comment (zeertzjq) [reply runtime/tutor/en/vim-01-beginner.tutor]: The `u` command is mentioned at the top of the tutor.

**Review by justinmk — APPROVED** [review]

**Comment by echometerain** [comment]
yay

**Comment by goyalyashpal** [comment]
Welll... undo... ewww... so all the checks stay broken even after finishing the test.  

and there's no feel good factor of seeing all the green happy checks.



the direction of add extra lines suggested initially was good.  sad it was ditched without giving its wording enough time & exposure to evolve.

**Comment by zeertzjq** [comment]
What's the difference? Both add the same number of extra lines.

**Comment by echometerain** [comment]
> Welll... undo... ewww... so all the checks stay broken even after finishing the test.
and there's no feel good factor of seeing all the green happy checks.

> What's the difference? Both add the same number of extra lines. If you add three empty lines you'll still see red checks.

I'm not sure what you guys are talking about.

The following is a comparison between adding extra lines and undoing. The check marks for subsequent tasks show up properly both times.

o
u

**Pull request merged**